substrate/frame/revive/src/evm/block_hash.rs

```diff
@@ -0,0 +1,873 @@
+const BLOOM_SIZE_BYTES: usize = 256;
+
+/// The intermediate representation of the Ethereum block builder.
+#[derive(Encode, Decode, TypeInfo)]
+pub struct EthereumBlockBuilderIR {
+	transaction_root_builder: Option<IncrementalHashBuilderIR>,
+	receipts_root_builder: Option<IncrementalHashBuilderIR>,
+
+	gas_used: U256,
+	pub(crate) tx_hashes: Vec<H256>,
+
+	logs_bloom: [u8; BLOOM_SIZE_BYTES],
+	pub(crate) gas_info: Vec<ReceiptGasInfo>,
+}
+
+impl Default for EthereumBlockBuilderIR {
+	fn default() -> Self {
+		Self {
+			transaction_root_builder: None,
+			receipts_root_builder: None,
+			gas_used: U256::zero(),
+			tx_hashes: Vec::new(),
+			logs_bloom: [0; BLOOM_SIZE_BYTES],
+			gas_info: Vec::new(),
+		}
+	}
+}
+
+/// Ethereum block builder.
+pub struct EthereumBlockBuilder {
+	pub(crate) transaction_root_builder: Option<IncrementalHashBuilder>,
+	pub(crate) receipts_root_builder: Option<IncrementalHashBuilder>,
+
+	gas_used: U256,
+	pub(crate) tx_hashes: Vec<H256>,
+
+	logs_bloom: LogsBloom,
+	gas_info: Vec<ReceiptGasInfo>,
+}
+
+impl EthereumBlockBuilder {
+	/// Constructs a new [`EthereumBlockBuilder`].
+	pub const fn new() -> Self {
+		Self {
+			transaction_root_builder: None,
+			receipts_root_builder: None,
+			gas_used: U256::zero(),
+			tx_hashes: Vec::new(),
+			logs_bloom: LogsBloom::new(),
+			gas_info: Vec::new(),
+		}
+	}
+
+	/// Converts the builder into an intermediate representation.
+	pub fn to_ir(self) -> EthereumBlockBuilderIR {
+		EthereumBlockBuilderIR {
+			transaction_root_builder: self.transaction_root_builder.map(|b| b.to_ir()),
+			receipts_root_builder: self.receipts_root_builder.map(|b| b.to_ir()),
+			gas_used: self.gas_used,
+			tx_hashes: self.tx_hashes,
+			logs_bloom: self.logs_bloom.bloom,
+			gas_info: self.gas_info,
+		}
+	}
+
+	/// Converts the intermediate representation back into a builder.
+	pub fn from_ir(ir: EthereumBlockBuilderIR) -> Self {
+		Self {
+			transaction_root_builder: ir
+				.transaction_root_builder
+				.map(|b| IncrementalHashBuilder::from_ir(b)),
+			receipts_root_builder: ir
+				.receipts_root_builder
+				.map(|b| IncrementalHashBuilder::from_ir(b)),
+			gas_used: ir.gas_used,
+			tx_hashes: ir.tx_hashes,
+			logs_bloom: LogsBloom { bloom: ir.logs_bloom },
+			gas_info: ir.gas_info,
+		}
+	}
+
+	/// Reset the state of the block builder to accommodate for the next block.
+	pub fn reset(&mut self) {
+		*self = Self::new();
+	}
+
+	/// Process a single transaction at a time.
+	pub fn process_transaction(
+		&mut self,
+		transaction_encoded: Vec<u8>,
+		success: bool,
+		gas_used: Weight,
+		encoded_logs: Vec<u8>,
+		receipt_bloom: LogsBloom,
+	) {
+		let tx_hash = H256(keccak_256(&transaction_encoded));
+		self.tx_hashes.push(tx_hash);
+
+		// Update the transaction trie.
+		let transaction_type = Self::extract_transaction_type(transaction_encoded.as_slice());
+		Self::add_builder_value(&mut self.transaction_root_builder, transaction_encoded);
```

Comment:
Just a random thought: Would is make sense to not store the first value in the incremental builder but just the tx index? And then when needed, we could load the data using the index from https://github.com/paritytech/polkadot-sdk/blob/6da298a9ace58e8b06f58f724ca9f974319db4cd/substrate/frame/system/src/lib.rs#L1023-L1025

It is just an idea that came to me, I expect it to be a bit lighter since the IR would lose the first value and we would need to serialize/deserialize less data. But not sure if it makes a big difference.